src/Processors/QueryPlan/IQueryPlanStep.h

```diff
@@ -113,6 +113,28 @@ class IQueryPlanStep
 
     virtual bool hasCorrelatedExpressions() const;
 
+
+    /// Returns true if the step has implemented removeUnusedColumns.
+    virtual bool canRemoveUnusedColumns() const { return false; }
+
+    struct UnusedColumnRemovalResult
+    {
+        bool updated_anything;
+        bool removed_any_input;
+    };
+
+    /// Removes the unnecessary inputs and outputs from the step based on required_outputs.
+    /// required_outputs must be a maybe empty subset of the current outputs of the step.
+    /// It is guaranteed that the output header of the step will contain all columns from
+    /// required_outputs and might contain some other columns too.
+    /// Can be used only if canRemoveUnusedColumns returns true.
+    /// If none of the output columns are removed, then it is preferable to keep the order
+    /// of output columns in order to avoid excessive updates of input headers in parent steps.
+    virtual UnusedColumnRemovalResult removeUnusedColumns(NameMultiSet /*required_outputs*/, bool /*remove_inputs*/);
```

Comment:
Possible later, but we can make this function position-based.
Get a list of output positions that are not needed, and return two lists of in/out positions that were removed.